Source/NSTextFinder.m

```diff
@@ -0,0 +1,305 @@
+*/
+
+#import <Foundation/NSArray.h>
+#import <Foundation/NSArchiver.h>
+
+#import "AppKit/NSTextFinder.h"
+#import "AppKit/NSTextView.h"
+
+#import "GSTextFinder.h"
+
+@implementation NSTextFinder
+
+- (instancetype) init
+{
+  self = [super init];
+  if (self != nil)
+    {
+      // initialize finder...
+      _finder = [GSTextFinder sharedTextFinder];
+    }
+  return self;
+}
+
+// Operation methods
+- (void) _showFindInterface
+{
+  [_finder showFindPanel];
+}
+
+- (void) _nextMatch
+{
+  [_finder findStringInTextView: (NSTextView *)_client
+                        forward: YES];
+}
+
+- (void) _previousMatch
+{
+  [_finder findStringInTextView: (NSTextView *)_client
+                        forward: NO];
+}
+
+- (void) _replaceAll
+{
+  [_finder replaceAllInTextView: (NSTextView *)_client
+                onlyInSelection: NO];
+}
+
+- (void) _replace
+{
+  [_finder replaceStringInTextView: (NSTextView *)_client];
+}
+
+- (void) _replaceAndFind
+{
+   [_finder replaceStringInTextView: (NSTextView *)_client];
+   [_finder findStringInTextView: (NSTextView *)_client
+                         forward: YES];
+}
+
+- (void) _setSearchString
+{
+  [_finder takeFindStringFromTextView: (NSTextView *)_client];
+}
+
+- (void) _replaceAllInSelection
+{
+  [_finder replaceAllInTextView: (NSTextView *)_client
+                onlyInSelection: YES];
+}
+
+- (void) _selectAll
+{
+}
+
+- (void) _selectAllInSelection
+{
+}
+
+- (void) _hideFindInterface
+{
+}
+
+- (void) _showReplaceInterface
+{
+  [_finder showFindPanel];
+}
+
+- (void) _hideReplaceInterface
+{
+}
+
+// Validating and performing
+- (void) performAction: (NSTextFinderAction)op
+{
+  switch (op)
+    {
+    case NSTextFinderActionShowFindInterface:
+      [self _showFindInterface];
+      break;
+    case NSTextFinderActionNextMatch:
+      [self _nextMatch];
+      break;
+    case NSTextFinderActionPreviousMatch:
+      [self _previousMatch];
+      break;
+    case NSTextFinderActionReplaceAll:
+      [self _replaceAll];
+      break;
+    case NSTextFinderActionReplace:
+      [self _replace];
+      break;
+    case NSTextFinderActionReplaceAndFind:
+      [self _replaceAndFind];
+      break;
+    case NSTextFinderActionSetSearchString:
+      [self _setSearchString];
+      break;
+    case NSTextFinderActionReplaceAllInSelection:
+      [self _replaceAllInSelection];
+      break;
+    case NSTextFinderActionSelectAll:
+      [self _selectAll];
+      break;
+    case NSTextFinderActionSelectAllInSelection:
+      [self _selectAllInSelection];
+      break;
+    case NSTextFinderActionHideFindInterface:
+      [self _hideFindInterface];
+      break;
+    case NSTextFinderActionShowReplaceInterface:
+      [self _showReplaceInterface];
+      break;
+    case NSTextFinderActionHideReplaceInterface:
+      [self _hideReplaceInterface];
+      break;
+    default:
+      NSLog(@"Unknown operation: %ld", op);
+      break;
+    }
+}
+
+- (BOOL) validateAction: (NSTextFinderAction)op
```

Comment:
The same here. We need to implement validateUserInterfaceItem: and call this method from there when the action is performTextFinderAction: or performFindPanelAction:.

Reply:
I don't know what the validateUserInterfaceItem method should look like.  Any suggestions?

Reply:
Done